DMP Filename handling

lpandath opened the pull request:
<!-- Please make sure you have read the submission guidelines before posting an PR -->
<!-- https://github.com/tuwien-csd/damap-backend/blob/next/CONTRIBUTING.md#pullrequests -->

## Description
The getDefaultFileName() method in DmpService was not handling null titles consistently, potentially resulting in filenames like "DMP_null_someID" for older DMPs.
Test cases for the getDefaultFileName() method were incomplete, missing scenarios for older DMPs with null titles and special characters

#### What does this PR do?
Always prefix filenames with "DMP_" for consistency
Handle null titles by using the DMP ID as a fallback
Added test case 

#### Breaking changes
<!-- Whether this PR contains breaking changes and which -->

#### Code review focus
<!-- What you want the reviewer to focus on -->

#### Dependencies
<!-- If this PR depends on or requires other/additional (frontend) changes, please list them here -->

### Checks
<!-- Adjust list as necessary -->
<!-- In case of DB changes make sure that names do not exceed 30 chars and that audit tables have been created/updated and do not contain FKs on entities. -->
- [ ] Tested with Oracle/PostgreSQL
- [ ] Export updated
- [ ] Documentation added
- [x] Tests added
- [ ] E2e tests created
- [ ] Successfully ran e2e tests before merge

closes GH-350

lpandath committed:
Fix: dmp export filename showing null

ValentinFutterer commented on src/main/java/org/damap/base/rest/dmp/service/DmpService.java:
```diff
@@ -233,11 +233,20 @@ public String getDefaultFileName(long id) {
                   + projectService.read(dmp.getProject().getUniversityId()).getAcronym()
                   + "_"
                   + formatter.format(date);
-        } else if (dmp.getProject().getTitle() != null)
+        } else if (dmp.getProject().getTitle() != null) {
           filename = "DMP_" + dmp.getProject().getTitle() + "_" + formatter.format(date);
+        } else {
+          filename = "DMP_" + id + "_" + formatter.format(date);
+        }
       } else {
-        if (dmp.getTitle() != null) filename = dmp.getTitle();
+        if (dmp.getTitle() != null) {
+          filename = dmp.getTitle();
+        } else {
+          filename = "DMP_" + id + "_" + formatter.format(date);
+        }
       }
+    } else {
+      filename = "DMP_" + id + "_" + formatter.format(date);
     }
 
```
This code works, but by adding more logic it became very hard to read. For example, there are three instances of:
```
else {
   filename = "DMP_" + id + "_" + formatter.format(date);
 }
```

Since this is the new default value, you could assign `filename = "DMP_" + id + "_" + formatter.format(date);` at the beginning, then you dont need to have all of these else branches. 
The you could turn this into an if else branch:
```
} else {
        if (dmp.getTitle() != null) {
          filename = dmp.getTitle();
```

Another thing of the top of my head would be something like this as a kind of preprocessing to avoid codelines that stretch over multiple lines.:
```
String universityId = dmp.getProject().getUniversityId();
Project universityProject = (universityId != null) ? projectService.read(universityId) : null;
```

I also noticed that when a project exists, but acronym and title are null, it will go to the default name, without even trying dmp.title().

lpandath replied:
Makes sense, thanks for your review!

ValentinFutterer commented:
Good work on the tests 👍

ValentinFutterer commented:
Seems like a git mishap has occured :)  can you please fix it, then I will continue the review

ValentinFutterer commented:
There is still something wrong with the git history - look at the changed files, they include changes from the last two patches.
Also, this change should only be one commit.

ValentinFutterer commented on src/main/java/org/damap/base/rest/dmp/service/DmpService.java:
```diff
@@ -212,37 +213,40 @@ private void removeVersions(Dmp dmp) {
-        } else if (dmp.getProject().getTitle() != null)
-          filename = "DMP_" + dmp.getProject().getTitle() + "_" + formatter.format(date);
-      } else {
-        if (dmp.getTitle() != null) filename = dmp.getTitle();
+    if (dmp == null) {
+      return filename;
+    }
+
+    if (dmp.getProject() != null) {
+      String universityId = dmp.getProject().getUniversityId();
+      ProjectDO universityProject =
+          (universityId != null) ? projectService.read(universityId) : null;
+
+      if (universityProject != null && universityProject.getAcronym() != null) {
+        return "DMP_" + universityProject.getAcronym() + "_" + formatter.format(date);
+      }
+
+      if (dmp.getProject().getTitle() != null) {
+        return "DMP_" + dmp.getProject().getTitle() + "_" + formatter.format(date);
       }
     }
 
-    filename = filename.replaceAll("[\"',\\s]+", "_");
+    if (dmp.getTitle() != null) {
+      return dmp.getTitle();
+    }
 
-    return filename;
+    return filename.replaceAll("[\"',\\s]+", "_");
```
There is some functionality lost from the refactor - the replaceAll postprocessing step is only done when no if is entered. You should use else ifs, to make the control flow only enter one conditional that sets the filename variable (instead of instantly returning). The last line here will then execute as normal and do the replaceAll.

ValentinFutterer replied:
Or, another idea, you scrap the global filename variable and do the postprocessing inline with the return statements, to improve readability of the if else constructs

ValentinFutterer commented on src/main/java/org/damap/base/rest/dmp/service/DmpService.java:
```diff
@@ -212,37 +213,56 @@ private void removeVersions(Dmp dmp) {
 
-    filename = filename.replaceAll("[\"',\\s]+", "_");
+      // This condition is met if:
+      // - no project (dmp.getProject() == null).
+      // - a project that contains neither an acronym nor a title
+      // - if dmp.getTitle() is available, we use it directly.
+      if (namePart == null && dmp.getTitle() != null) {
+        namePart = dmp.getTitle();
+        usePrefixAndSuffix = false;
+      }
+    }
 
-    return filename;
+    String finalName;
+    if (namePart != null) {
+      if (usePrefixAndSuffix) {
+        finalName = "DMP_" + namePart + "_" + formattedDate;
+      } else {
+        finalName = namePart;
+      }
+    } else {
+      // Fallback to default if no namePart could be determined from DMP or Project
+      finalName = "DMP_" + id + "_" + formattedDate;
+    }
```
This part is weird. `usePrefixAndSuffix` can only be false if `namePart = null`. But `usePrefixAndSuffix` only matters in a block that is entered when `namePart != null`...

Could you please fix this? Otherwise, I will do it myself. I have already spent more time reviewing than it would take me to solve this issue myself.

lpandath replied:
@ValentinFutterer thank you, i will check why is weird. What is your opinion on this as reviwer? otherwise i will come up with different solution

ValentinFutterer replied:
Its weird to read because when namepart is null, its set to the dmp title and the `usePrefixAndSuffix`. And then in the next block, where it uses `usePrefixAndSuffix` as a condition, it checks if `namePart != null`. And i missed the logical connection where namePart was set to the title, so its not null and the block is entered. You need to make a lot of logical jumps to understand it.

lpandath replied:
replied in the other comment.

lpandath committed:
fix: resolve logical contradiction in DMP filename

- removing the namePart == null check when setting usePrefixAndSuffix
- making the flag properly tied to the source of the name - DMP title

ValentinFutterer commented on src/main/java/org/damap/base/rest/dmp/service/DmpService.java:
```diff
@@ -212,37 +213,48 @@ private void removeVersions(Dmp dmp) {
+      finalName = usePrefixAndSuffix ? "DMP_" + namePart + "_" + formattedDate : namePart;
+    } else {
+      finalName = "DMP_" + id + "_" + formattedDate;
+    }
```
Problem with this is, that namePart is now always overriden. The original solution was way to o complicated to be easily understood. I am not a big fan of flags like `useprefixAndSuffix`, especially when its set at only one place - its just one more thing to track. The original intention was to handle a weird exception to the naming algorithm, but why use a flag at the first place. Something like this would be easier too read imo:

```
String finalName;
    if (namePart == null) {
      if (dmp != null && dmp.getTitle() != null) {
        finalName = dmp.getTitle();
      } else {
        finalName = "DMP_" + id + "_" + formattedDate;
      }
    } else {
      finalName = "DMP_" + namePart + "_" + formattedDate;
    }
```

But thats still a lot of code and if and elses. To make it even cleaner, we could just treat `dmp.getTitle()` like a normal namePart with prefix and suffix. To do this, we can just replace all empty spaces with underscores. What do you think of that idea, would that work for you and Graz?

lpandath replied:
Thank you for your thorough review! Let me address this:
the current implementation has some merits - it provides a clear separation between name gathering and name formatting, with the namePart variable acting as a single point of truth for the filename's base. The usePrefixAndSuffix flag explicitly documents our rule where DMP titles are treated differently.

However, you are right about the readability issues. The code evolved from a simpler version into a complex structure with nested if-else statements, which is exactly why I created this issue and attempted to refactor it.

I'll restructure the code to remove the usePrefixAndSuffix flag and handle each case separately with more clear methods where this will make the priority order of name selection while maintaining the same rules but with cleaner code.

p.s `squash` after review.

lpandath committed:
refactor: improve DMP filename generation readability

ValentinFutterer commented:
It works fine now and the code is way more readable 👍 
The banch is out of date though, can you rebase and squash, then i can merge.

lpandath commented:
> It works fine now and the code is way more readable 👍 The banch is out of date though, can you rebase and squash, then i can merge.

you can do that now, thanks!

bitmotivator commented on src/main/java/org/damap/base/rest/dmp/service/DmpService.java:
```diff
@@ -212,37 +212,61 @@ private void removeVersions(Dmp dmp) {
+
+  private String extractProjectName(Dmp dmp) {
+    if (dmp.getProject() == null) {
+      return null;
```
In the old code this would have created a filename based on `My Data Management Plan`. The new code seems to create an artificial filename based on the DMP date. Is this intentional?

lpandath replied:
The old implementation had issues with older DMPs, were resulting in filenames like "DMP_null_someID" for older DMPs. The "My Data Management Plan" filename is now only used as the ultimate fallback when the DMP doesn't exist in the database. For existing DMPs, users get more detailed filenames based on their actual DMP title or project information with date...

bitmotivator replied:
I have no strong opinion on file names, maybe @ValentinFutterer does, I just wanted to note that the logic seems to be different.

ValentinFutterer replied:
I also dont have strong opinions, these are edge cases that shouldnt even come up in normal deployments and users dont pay that much attention to these filenames anyways. I just want to merge this PR already 😅

bitmotivator reviewed (approved):
The old code took me a while to wrap my head around, but the new code looks fine. @ValentinFutterer please take a second look and then I'm happy to merge this.
